libyul/AsmJsonConverter.cpp

```diff
@@ -170,27 +172,55 @@ Json::Value AsmJsonConverter::operator()(ForLoop const& _node) const
 
 Json::Value AsmJsonConverter::operator()(Break const& _node) const
 {
-	return createAstNode(nativeLocationOf(_node), "YulBreak");
+	return createAstNode(debugDataOf(_node).get(), "YulBreak");
 }
 
 Json::Value AsmJsonConverter::operator()(Continue const& _node) const
 {
-	return createAstNode(nativeLocationOf(_node), "YulContinue");
+	return createAstNode(debugDataOf(_node).get(), "YulContinue");
 }
 
 Json::Value AsmJsonConverter::operator()(Leave const& _node) const
 {
-	return createAstNode(nativeLocationOf(_node), "YulLeave");
+	return createAstNode(debugDataOf(_node).get(), "YulLeave");
 }
 
-Json::Value AsmJsonConverter::createAstNode(langutil::SourceLocation const& _location, string _nodeType) const
+string AsmJsonConverter::formatLocationWithFileName(SourceLocation const& _location)
 {
+	return
+		to_string(_location.start) +
+		":" +
+		to_string(_location.length()) +
+		":" +
+		(_location.sourceName ? *_location.sourceName : "");
+}
+
+string AsmJsonConverter::formatLocationWithFileIndex(SourceLocation const& _location, optional<size_t> const& _sourceIndex)
+{
+	return
+		to_string(_location.start) +
+		":" +
+		to_string(_location.length()) +
+		":" +
+		(_sourceIndex.has_value() ? to_string(_sourceIndex.value()) : "-1");
+}
+
+Json::Value AsmJsonConverter::createAstNode(DebugData const& _debugData, string _nodeType) const
+{
+	yulAssert(_debugData, "");
+
 	Json::Value ret{Json::objectValue};
 	ret["nodeType"] = std::move(_nodeType);
-	int length = -1;
-	if (_location.start >= 0 && _location.end >= 0)
-		length = _location.end - _location.start;
-	ret["src"] = to_string(_location.start) + ":" + to_string(length) + ":" + (m_sourceIndex.has_value() ? to_string(m_sourceIndex.value()) : "-1");
+
+	ret["src"] = formatLocationWithFileIndex(_debugData->nativeLocation, m_sourceIndex);
+	if (_debugData->originLocation.isValid())
+	{
+		if (_debugData->nativeLocation.sourceName == _debugData->originLocation.sourceName)
+			ret["origin"] = formatLocationWithFileIndex(_debugData->originLocation, m_sourceIndex);
+		else
+			ret["origin"] = formatLocationWithFileName(_debugData->originLocation);
+	}
```

Comment:
The problem with this PR is that we don't have indexes for the original Solidity files when compiling Yul.

This does work because currently we do not support outputting Yul AST (#9590) but will break when we implement it so I don't think this is a good solution.